### PR TITLE
trillian/ctfe: add misbehaving log mode

### PR DESCRIPTION
**Pull request by daviddrysdale**
Only intended for testing, should never be turned on.

Add a variety of misbehaviours in the CTFE's behaviour, each
occuring with a 1% chance.

### CODE REVIEW

**Comment by daviddrysdale** [comment]
This would be too scary to merge, so I'm suggesting keeping it as a separate branch forever, occasionally merging in current `master` so it's possible to manually build a bad CTFE.

**Comment by Martin2112** [comment]
Can throw a couple of alternative ideas out. Won't be offended if you tell me they're no good.

1.) Could have a separate binary in a different package making it impossible to set this flag on the normal one.
2.) Could have a wrapper round handlers maybe, move that to the same place as the above and it will gain more safety.
3.) We could write a proxy that messes with log responses, then it would be entirely separate but perhaps not as flexible.

**Comment by daviddrysdale** [comment]
All sensible ideas, but I'm generally trying to keep this minimum-effort -- I think the alternatives all involve more work (but have the advantage that there would be no manual merge effort).  Other thoughts:

1) Even if the flag can't be set on the normal `ct_server` binary, I'd still be nervous about the mere existence of the misbehaviour code in `handlers.go`.
2) Helps with that concern, but it either involves more work (to un-parse, break and then re-parse&sign things) or has more limited scope (like 3)
3) As you point out, a proxy would be unable to produce several of the failure cases that are included here.

**Comment by Martin2112** [comment]
A proxy could be given the same test key but might still not be flexible enough.

**Comment by daviddrysdale** [comment]
I think a proxy with the key would be able to misbehave in all the same ways, but it's just a bunch more effort.

File: trillian/ctfe/handlers.go [hunk 1]
```diff
@@ -456,13 +458,35 @@ func addChainInternal(ctx context.Context, li *logInfo, w http.ResponseWriter, r
 	} else if len(rest) > 0 {
 		return http.StatusInternalServerError, fmt.Errorf("extra data (%d bytes) on reconstructing MerkleTreeLeaf", len(rest))
 	}
+	sctVersion := ct.V1
+	if li.misbehave("use submitted leaf not already-logged leaf (if present) in SCT") {
```
Comment (AlCutter) [inline trillian/ctfe/handlers.go]: Does this imply a n*(1/2, 1/4, 1/8) probability across these types of misbehavior? (passim)
 
maybe  something like
```golang
type misbehaviour struct {
  desc string
  playup func()
}

func (li *...) chooseToBeBad(behaviours []behaviour) {
  if !badlog {return}

  if rand(100) != 0 { return } 
  b := misbehaviours[rand(len(misbehaviours))]
  log.INFO("DELIBERATE: %s", b.desc)
  b.playUp()
}


...
   ...
   li.chooseToBeBad([]behaviours{
     { "use submitted leaf..", func() { ... } },
     { "change SCT timestamp...", func() { ... }}.
     ...
      {...})
```

or similar?

**Comment by AlCutter** [comment]
I share the concern about checking this code in, but I don't really have any better suggestions at the mo'...

File: trillian/ctfe/handlers.go (hunk 1 above)
Comment (daviddrysdale) [reply trillian/ctfe/handlers.go]: It does imply a reduced chance for later misbehaviour, but it's a (0.99)<sup>N-1</sup>/100 chance for the N-th arm:
 - first arm has a 0.01 (=1/100) chance of happening
 - second arm has a 0.0099 (=99/100 * 1/100) chance
 - third arm has a 0.009801 (=99/100 * 99/100 * 1/100) chance
 - fourth arm has a 0.009703 (=99/100 * 99/100 * 99/100 * 1/100) chance
 - fifth arm has a 0.009606 (=99/100 * 99/100 * 99/100 * 99/100 * 1/100) chance.

So with a fixed misbehaviour rate of 1% I don't think the scaling is worth worring about.  If the rate were configurable, then I agree that something more sophisticated would be worthwhile, but I'm generally trying to keep this set of diffs as simple as possible (given the intent of maintaining a separate branch).

**Comment by daviddrysdale** [comment]
I think the best alternative is a proxy that has a private key (either its own or a copy of the logs) so that it can make arbitrary manipulations.  But that would be a lot more work than this.

**Comment by daviddrysdale** [comment]
I've pushed the `bad-to-the-bone` branch to the main repo so it's available for future use regardless of whether I accidentally press the 'Delete Fork' button on my repo...

**Comment by daviddrysdale** [comment]
This code has been pushed to the main repo as a separate `bad-to-the-bone` branch, so it is available for experimentation and occasional rebasing, but isn't in `master` to reduce risk.

**Pull request closed without merging**